tools/querytee/goldfish/storage_cloudsql.go

```diff
@@ -0,0 +1,230 @@
+package goldfish
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	_ "github.com/go-sql-driver/mysql" // MySQL driver
+)
+
+// CloudSQLStorage implements Storage for Google Cloud SQL
+type CloudSQLStorage struct {
+	db     *sql.DB
+	config StorageConfig
+}
+
+// NewCloudSQLStorage creates a new CloudSQL storage backend
+func NewCloudSQLStorage(config StorageConfig, password string) (*CloudSQLStorage, error) {
+	if password == "" {
+		return nil, fmt.Errorf("CloudSQL password must be provided via GOLDFISH_DB_PASSWORD environment variable")
+	}
+
+	// Build DSN for CloudSQL proxy connection
+	// MySQL DSN format: username:password@tcp(host:port)/dbname?params
+	dsn := fmt.Sprintf("%s:%s@tcp(%s:%d)/%s?parseTime=true&charset=utf8mb4&collation=utf8mb4_unicode_ci",
+		config.CloudSQLUser,
+		password,
+		config.CloudSQLHost,
+		config.CloudSQLPort,
+		config.CloudSQLDatabase,
+	)
+
+	db, err := sql.Open("mysql", dsn)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open database: %w", err)
+	}
+
+	// Configure connection pool
+	db.SetMaxOpenConns(config.MaxConnections)
+	db.SetMaxIdleConns(config.MaxConnections / 2)
+	db.SetConnMaxIdleTime(time.Duration(config.MaxIdleTime) * time.Second)
+
+	// Verify connection
+	if err := db.Ping(); err != nil {
+		db.Close()
+		return nil, fmt.Errorf("failed to ping database: %w", err)
+	}
+
+	// Initialize schema
+	if err := initSchema(db); err != nil {
+		db.Close()
+		return nil, fmt.Errorf("failed to initialize schema: %w", err)
+	}
+
+	return &CloudSQLStorage{
+		db:     db,
+		config: config,
+	}, nil
+}
+
+// StoreQuerySample stores a sampled query with performance statistics
+func (s *CloudSQLStorage) StoreQuerySample(ctx context.Context, sample *QuerySample) error {
+	query := `
+		INSERT INTO sampled_queries (
+			correlation_id, tenant_id, query, query_type,
+			start_time, end_time, step_duration,
+			cell_a_exec_time_ms, cell_b_exec_time_ms,
+			cell_a_queue_time_ms, cell_b_queue_time_ms,
+			cell_a_bytes_processed, cell_b_bytes_processed,
+			cell_a_lines_processed, cell_b_lines_processed,
+			cell_a_bytes_per_second, cell_b_bytes_per_second,
+			cell_a_lines_per_second, cell_b_lines_per_second,
+			cell_a_entries_returned, cell_b_entries_returned,
+			cell_a_splits, cell_b_splits,
+			cell_a_shards, cell_b_shards,
+			cell_a_response_hash, cell_b_response_hash,
+			cell_a_response_size, cell_b_response_size,
+			cell_a_status_code, cell_b_status_code,
+			sampled_at
+		) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
+	`
+
+	_, err := s.db.ExecContext(ctx, query,
```

Comment:
do we need to do any sanitization here? for example tenant ID comes from a header, so could easily be injectable.